clippy_utils/src/check_proc_macro.rs

```diff
@@ -250,6 +250,27 @@ fn variant_search_pat(v: &Variant<'_>) -> (Pat, Pat) {
     }
 }
 
+fn fn_kind_pat(tcx: TyCtxt<'_>, kind: &FnKind<'_>, body: &Body<'_>, hir_id: HirId) -> (Pat, Pat) {
+    let (start_pat, end_pat, visibility) = match kind {
+        FnKind::ItemFn(.., header) => (
+            fn_header_search_pat(*header),
+            Pat::Str(""),
+            tcx.visibility(tcx.hir().local_def_id(hir_id)),
+        ),
+        FnKind::Method(.., sig) => (
+            fn_header_search_pat(sig.header),
+            Pat::Str(""),
+            tcx.visibility(tcx.hir().local_def_id(hir_id)),
+        ),
+        FnKind::Closure => return (Pat::Str(""), expr_search_pat(tcx, &body.value).1),
+    };
+    if visibility.is_public() {
+        (Pat::Str("pub"), end_pat)
+    } else {
+        (start_pat, end_pat)
+    }
```

Comment:
Not entirely sure what is needed to be done here, do I have to cover more visibility kinds?